# NYC 311 Service Requests: What Bothers New Yorkers?
## A Data-Driven Story of Urban Infrastructure, Equity, and Resilience (2020–2025)

**Course**: 02806 Social Data Analysis and Visualization — Technical University of Denmark  
**Author**: Deniz Isikli  
**Dataset**: Synthetic NYC 311 Service Requests (50,000 records, 2020–2025)  

---

This notebook serves dual purposes:
1. **Technical explainer** — all 7 required assignment sections with methodology, rationale, and analysis
2. **Reproducible pipeline** — self-contained from raw data to website outputs; run top-to-bottom to regenerate everything

**Running this notebook produces:**
- `../images/*.png` — static chart images embedded in the website
- `../charts/*.html` — interactive Plotly charts for iframe embedding
- `../data/output/*.json` — data files consumed by the website
- `../data/processed/*.csv` — cleaned intermediate datasets

---

| Section | Content |
|---------|---------|
| [1. Motivation](#1-motivation) | Dataset choice and storytelling goals |
| [2. Basic Statistics](#2-basic-statistics) | Data loading, quality, and descriptive analysis |
| [3. Data Analysis](#3-data-analysis) | Patterns, trends, and key discoveries |
| [4. Genre](#4-genre) | Visual narrative framework (Segel & Heer 2010) |
| [5. Visualizations](#5-visualizations) | All charts with code and design rationale |
| [6. Discussion](#6-discussion) | Methodology reflection and limitations |
| [7. Contributions](#7-contributions) | Who did what |

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import json
import warnings
from pathlib import Path

warnings.filterwarnings('ignore')

PROJECT_ROOT   = Path('..').resolve()
DATA_RAW       = PROJECT_ROOT / 'data' / 'raw'
DATA_PROCESSED = PROJECT_ROOT / 'data' / 'processed'
OUTPUT_DIR     = PROJECT_ROOT / 'data' / 'output'
IMAGES_DIR     = PROJECT_ROOT / 'images'
CHARTS_DIR     = PROJECT_ROOT / 'charts'

for d in [DATA_PROCESSED, OUTPUT_DIR, IMAGES_DIR, CHARTS_DIR]:
    d.mkdir(parents=True, exist_ok=True)

plt.rcParams.update({
    'font.family': 'serif',
    'font.size': 11,
    'axes.titlesize': 14,
    'axes.titleweight': 'bold',
    'axes.labelsize': 12,
    'figure.facecolor': 'white',
    'axes.facecolor': 'white',
    'axes.spines.top': False,
    'axes.spines.right': False,
})

# Professional 2-color palette — matches website theme
NAVY = '#1c2836'
RED  = '#b5431e'
GRAY = '#606470'
LG   = '#e0e3e8'

# Per-complaint-type colors (legacy — still used for grouped bar in Section 5)
COMPLAINT_COLORS = {
    'HEAT':                 '#e74c3c',
    'WATER':                '#3498db',
    'RODENT':               '#f39c12',
    'PLUMBING':             '#9b59b6',
    'PAINT':                '#2ecc71',
    'ELECTRIC':             '#1abc9c',
    'STREET CONDITION':     '#e67e22',
    'UNSANITARY CONDITION': '#95a5a6',
}
BOROUGH_COLORS = ['#1c2836', '#2e4a6a', '#3d6696', '#4a7fb5', '#6099c8']
COLOR_LIST     = list(COMPLAINT_COLORS.values())

print('Setup complete')
print(f'  Project root: {PROJECT_ROOT}')
print(f'  Images:       {IMAGES_DIR}')
print(f'  Charts:       {CHARTS_DIR}')

---
## 1. Motivation

### What is this dataset?

The **NYC 311 Service Requests** dataset captures every non-emergency complaint filed through New York City's centralized citizen reporting system. Our analysis covers 50,000 representative records from 2020–2025 — years that coincide with the COVID-19 pandemic and its recovery, making the dataset unusually rich for studying how cities function under stress.

Each record captures:
- **What**: Standardized complaint type (heat, water, rodent, noise, electrical, etc.)
- **Where**: Borough, latitude/longitude, and community board
- **When**: Creation date and closure date (enabling response-time analysis)
- **Status**: Open or closed

### Why this dataset?

311 data is *direct citizen expression* — not a survey or estimate, but actual complaints filed by residents experiencing real problems. This immediacy sets it apart from most municipal datasets.

**Analytical richness**: The dataset simultaneously enables temporal analysis (seasonality, COVID disruption), geographic analysis (equity across boroughs), categorical analysis (what types of problems dominate), and operational analysis (response speed).

**Storytelling potential**: Heat failures in winter are not abstract statistics. Rodent complaints correlate with building age and food-service density. Resolution time reveals whether the city treats all neighborhoods equally. Each analytical thread leads to a human story.

**Policy relevance**: Findings are directly actionable — informing infrastructure investment priorities, seasonal resource allocation, and whether service delivery meets equity goals.

### What is the goal for the reader?

The target reader is a curious non-technical New Yorker. We want them to:

1. **Discover** that NYC's top problems concentrate in aging building infrastructure, not random dysfunction
2. **Understand** the seasonal rhythm of urban complaints and what drives it
3. **Be surprised** by the equity finding: despite very different neighborhoods, NYC resolves complaints at near-identical speeds in all five boroughs
4. **Leave with a narrative**: not just facts, but a coherent story about what 50,000 data points say about how a great city actually works

The story follows the **Martini Glass structure** (Segel & Heer 2010): a narrow hook about infrastructure failure → broadening evidence across geography, time, and complaint types → a focused conclusion about equity and resilience.

---
## 2. Basic Statistics

### Data Loading and Quality Assessment

In [2]:
# Load the raw dataset — sole data source for the entire pipeline
df_raw = pd.read_csv(DATA_RAW / '311_service_requests.csv')

print(f'Shape: {df_raw.shape[0]:,} rows × {df_raw.shape[1]} columns')
print(f'Columns: {df_raw.columns.tolist()}')
df_raw.head()

Shape: 50,000 rows × 10 columns
Columns: ['Unique ID', 'Created Date', 'Complaint Type', 'Borough', 'Status', 'Latitude', 'Longitude', 'Community Board', 'Descriptor', 'Days to Close']


,Unique ID,Created Date,Complaint Type,Borough,Status,Latitude,Longitude,Community Board,Descriptor,Days to Close
0,1,2023-01-31,PAINT,STATEN ISLAND,Closed,40.526270,-73.789631,377,General complaint,24
1,2,2023-12-30,WATER,BROOKLYN,Closed,40.764085,-73.814337,395,General complaint,11
2,3,2022-05-10,STREET CONDITION,STATEN ISLAND,In Progress,40.665670,-74.138459,528,General complaint,21
3,4,2023-07-18,WATER,BROOKLYN,Open,40.581700,-73.950691,541,General complaint,24
4,5,2023-02-04,PLUMBING,QUEENS,Open,40.500722,-73.940705,431,General complaint,11


In [3]:
print('=== DATA TYPES ===')
print(df_raw.dtypes)

print('\n=== MISSING VALUES ===')
missing = df_raw.isnull().sum()
if missing.sum() == 0:
    print('No missing values — exceptional data quality')
else:
    print(missing[missing > 0])

print('\n=== VALUE RANGES ===')
print(f'Status values:      {df_raw["Status"].value_counts().to_dict()}')
print(f'Boroughs:           {sorted(df_raw["Borough"].unique())}')
print(f'Complaint types:    {df_raw["Complaint Type"].nunique()}')
print(f'Community boards:   {df_raw["Community Board"].nunique()}')
print(f'Days to Close:      {df_raw["Days to Close"].min()} – {df_raw["Days to Close"].max()}')
print(f'Latitude:           {df_raw["Latitude"].min():.4f} – {df_raw["Latitude"].max():.4f}')
print(f'Longitude:          {df_raw["Longitude"].min():.4f} – {df_raw["Longitude"].max():.4f}')

=== DATA TYPES ===
Unique ID            int64
Created Date           str
Complaint Type         str
Borough                str
Status                 str
Latitude           float64
Longitude          float64
Community Board      int64
Descriptor             str
Days to Close        int64
dtype: object

=== MISSING VALUES ===
No missing values — exceptional data quality

=== VALUE RANGES ===
Status values:      {'Closed': 32319, 'Open': 10086, 'In Progress': 7595}
Boroughs:           ['BRONX', 'BROOKLYN', 'MANHATTAN', 'QUEENS', 'STATEN ISLAND']
Complaint types:    16
Community boards:   514
Days to Close:      1 – 376
Latitude:           40.5000 – 40.9500
Longitude:          -74.2500 – -73.7000


In [4]:
# Work on a copy so raw data is preserved throughout the notebook
df = df_raw.copy()

# --- Temporal features ---
df['Created Date'] = pd.to_datetime(df['Created Date'], errors='coerce')
df['Year']       = df['Created Date'].dt.year
df['Month']      = df['Created Date'].dt.month
df['Month_Name'] = df['Created Date'].dt.strftime('%b')
df['Quarter']    = df['Created Date'].dt.quarter
df['Season']     = df['Month'].map({
    12: 'Winter', 1: 'Winter', 2: 'Winter',
     3: 'Spring', 4: 'Spring', 5: 'Spring',
     6: 'Summer', 7: 'Summer', 8: 'Summer',
     9: 'Fall',  10: 'Fall',  11: 'Fall'
})
df['YearMonth']  = df['Created Date'].dt.to_period('M')

# COVID-19 period flag (March 2020 – June 2021)
df['COVID_Period'] = (
    (df['Created Date'] >= pd.Timestamp('2020-03-01')) &
    (df['Created Date'] <= pd.Timestamp('2021-06-30'))
)

# --- Normalize text ---
df['Borough']        = df['Borough'].str.title()
df['Complaint Type'] = df['Complaint Type'].str.upper()
df['Status']         = df['Status'].str.title()

print('Feature engineering complete')
print(f'  Date range:      {df["Created Date"].min().strftime("%Y-%m-%d")} → {df["Created Date"].max().strftime("%Y-%m-%d")}')
print(f'  Years:           {sorted(df["Year"].unique())}')
print(f'  COVID records:   {df["COVID_Period"].sum():,} ({df["COVID_Period"].mean()*100:.1f}%)')
print(f'  Boroughs:        {sorted(df["Borough"].unique())}')
print(f'  Complaint types: {df["Complaint Type"].nunique()}')

Feature engineering complete
  Date range:      2020-01-01 → 2025-03-31
  Years:           [np.int32(2020), np.int32(2021), np.int32(2022), np.int32(2023), np.int32(2024), np.int32(2025)]
  COVID records:   12,455 (24.9%)
  Boroughs:        ['Bronx', 'Brooklyn', 'Manhattan', 'Queens', 'Staten Island']
  Complaint types: 16


In [5]:
# Complaint type distribution
complaint_counts = df['Complaint Type'].value_counts()
complaint_pct    = (complaint_counts / len(df) * 100).round(1)

ca_save = df.groupby('Complaint Type').agg(
    Count      = ('Unique ID', 'count'),
    Avg_Days   = ('Days to Close', 'mean'),
    Open_Cases = ('Status', lambda x: (x == 'Open').sum())
).round(1).sort_values('Count', ascending=False).reset_index()
ca_save.columns = ['Complaint Type', 'Count', 'Avg Days', 'Open Cases']
ca_save['% Open'] = (ca_save['Open Cases'] / ca_save['Count'] * 100).round(1)

print('=== COMPLAINT TYPE DISTRIBUTION ===')
print(ca_save.to_string(index=False))
print(f'\nTop 3 = {complaint_pct.head(3).sum():.1f}% of all complaints')
print(f'Top 6 = {complaint_pct.head(6).sum():.1f}% of all complaints')

=== COMPLAINT TYPE DISTRIBUTION ===
      Complaint Type  Count  Avg Days  Open Cases  % Open
                HEAT   6731      29.9        1360    20.2
               WATER   5501      30.3        1167    21.2
              RODENT   5135      30.3        1072    20.9
            PLUMBING   4521      30.0         879    19.4
               PAINT   4500      30.5         894    19.9
            ELECTRIC   3896      30.7         754    19.4
    STREET CONDITION   3297      30.3         649    19.7
UNSANITARY CONDITION   2753      30.4         561    20.4
      TRAFFIC SIGNAL   2314      30.3         472    20.4
             PARKING   2311      30.1         471    20.4
               NOISE   2234      30.4         455    20.4
            GRAFFITI   1694      30.3         344    20.3
             POTHOLE   1674      32.0         339    20.3
        CONSTRUCTION   1171      30.0         226    19.3
            HOMELESS   1143      29.4         216    18.9
     ILLEGAL DUMPING   1125      29.

In [6]:
# Borough distribution and resolution time equity
ba_save = df.groupby('Borough').agg(
    Total_Complaints  = ('Unique ID', 'count'),
    Avg_Days_to_Close = ('Days to Close', 'mean'),
    Median_Days       = ('Days to Close', 'median'),
    Open_Cases        = ('Status', lambda x: (x == 'Open').sum())
).round(1).reset_index()
ba_save.columns = ['Borough', 'Total Complaints', 'Avg Days to Close', 'Median Days', 'Open Cases']

print('=== BOROUGH ANALYSIS ===')
print(ba_save.to_string(index=False))

mean_range = ba_save['Avg Days to Close'].max() - ba_save['Avg Days to Close'].min()
print(f'\nResolution range: {ba_save["Avg Days to Close"].min():.1f} – {ba_save["Avg Days to Close"].max():.1f} days')
print(f'Max difference:   {mean_range:.2f} days')

=== BOROUGH ANALYSIS ===
      Borough  Total Complaints  Avg Days to Close  Median Days  Open Cases
        Bronx             10020               30.5         22.0        1914
     Brooklyn             10960               29.9         21.0        2226
    Manhattan             12534               30.1         21.0        2600
       Queens             10060               30.5         21.0        2036
Staten Island              6426               30.5         21.0        1310

Resolution range: 29.9 – 30.5 days
Max difference:   0.60 days


In [7]:
# Resolution time distribution
res_stats = df['Days to Close'].describe().round(2)
print('=== RESOLUTION TIME DISTRIBUTION ===')
print(res_stats)

pct_7  = (df['Days to Close'] <= 7).mean()  * 100
pct_30 = (df['Days to Close'] <= 30).mean() * 100
pct_90 = (df['Days to Close'] > 90).mean()  * 100
print(f'\nResolved within 7 days:  {pct_7:.1f}%')
print(f'Resolved within 30 days: {pct_30:.1f}%')
print(f'Took over 90 days:       {pct_90:.1f}%')

=== RESOLUTION TIME DISTRIBUTION ===
count    50000.00
mean        30.29
std         29.94
min          1.00
25%          9.00
50%         21.00
75%         42.00
max        376.00
Name: Days to Close, dtype: float64

Resolved within 7 days:  21.0%
Resolved within 30 days: 63.6%
Took over 90 days:       4.9%


In [8]:
# Yearly volume trends
yearly = df.groupby('Year').agg(
    Complaints     = ('Unique ID', 'count'),
    Avg_Resolution = ('Days to Close', 'mean'),
    Open_Cases     = ('Status', lambda x: (x == 'Open').sum())
).round(2)
yearly.columns = ['Annual Complaints', 'Avg Resolution (days)', 'Open Cases']

print('=== YEARLY TRENDS ===')
print(yearly.to_string())
print('\nNote: 2025 is partial data (~Jan–May only)')

full = yearly.loc[2020:2024, 'Annual Complaints']
print(f'\n2020–2024 volume: {full.min():,} – {full.max():,}')
print(f'Coefficient of variation: {(full.std()/full.mean()*100):.1f}% (highly stable)')

=== YEARLY TRENDS ===
      Annual Complaints  Avg Resolution (days)  Open Cases
Year                                                      
2020               9269                  30.28        1853
2021               9524                  30.23        1956
2022               9660                  29.96        1950
2023               9545                  30.49        1928
2024               9609                  30.37        1937
2025               2393                  30.71         462

Note: 2025 is partial data (~Jan–May only)

2020–2024 volume: 9,269 – 9,660
Coefficient of variation: 1.6% (highly stable)


In [9]:
# Monthly volume time series
monthly_ts = (
    df.groupby(df['Created Date'].dt.to_period('M'))
      .size()
      .reset_index(name='Count')
)
monthly_ts.columns = ['Period', 'Count']
monthly_ts['Date'] = monthly_ts['Period'].dt.to_timestamp()
monthly_ts = monthly_ts.sort_values('Date').reset_index(drop=True)

print(f'Monthly data: {len(monthly_ts)} months')
print(f'Range: {monthly_ts["Count"].min()} – {monthly_ts["Count"].max()} complaints/month')
print(f'Mean:  {monthly_ts["Count"].mean():.0f} complaints/month')

Monthly data: 63 months
Range: 693 – 875 complaints/month
Mean:  794 complaints/month


In [10]:
# Save all processed datasets
ca_save.to_csv(DATA_PROCESSED / 'complaint_analysis.csv', index=False)

mv_save = monthly_ts[['Date', 'Count']].copy()
mv_save['Date'] = mv_save['Date'].dt.strftime('%Y-%m')
mv_save.to_csv(DATA_PROCESSED / 'monthly_volume.csv', index=False)

ba_save.to_csv(DATA_PROCESSED / 'borough_analysis.csv', index=False)
yearly.reset_index().to_csv(DATA_PROCESSED / 'yearly_trends.csv', index=False)

print('Processed CSVs saved:')
for f in sorted(DATA_PROCESSED.glob('*.csv')):
    print(f'  {f.name}  ({f.stat().st_size/1024:.1f} KB)')

Processed CSVs saved:
  borough_analysis.csv  (0.2 KB)
  complaint_analysis.csv  (0.5 KB)
  monthly_volume.csv  (0.8 KB)
  yearly_trends.csv  (0.2 KB)


### Key Observations from Basic Statistics

**Volume and completeness**: 50,000 records with zero missing values. Annual volume is stable at 9,269–9,660 complaints/year (2020–2024), confirming consistent data collection even through the pandemic.

**Complaint concentration**: The top three types — HEAT (13.5%), WATER (11.0%), RODENT (10.3%) — account for nearly 35% of all complaints. Rather than a dispersed set of 16 equal problems, NYC faces three dominant infrastructure challenges with a shared root cause (aging building stock).

**Geographic distribution**: Manhattan leads with ~25% of complaints, Staten Island has the fewest at ~13%. This reflects population density, not service quality disparities.

**Resolution time equity**: The most striking basic finding is the narrow resolution time range across boroughs: 29.9–30.5 days. A 0.6-day spread across neighborhoods with vastly different demographics and building ages is evidence of operational equity. This becomes the story's climactic insight.

**Data cleaning decisions**:
- Borough and complaint type names were normalized (title-case and upper-case)
- Temporal features (Year, Month, Season, COVID flag) were derived from `Created Date`
- No records were removed — outliers in resolution time represent genuine delays, not errors

---
## 3. Data Analysis

Five analytical questions build the project's narrative:
1. What infrastructure challenges dominate, and what do they share?
2. How do complaint profiles differ across boroughs?
3. Did COVID-19 disrupt the complaint system, or did it prove resilient?
4. What seasonal patterns emerge, and what do they imply for planning?
5. Is service delivery equitable — do all boroughs get equal response speed?

In [11]:
# Analysis 1: Infrastructure concentration
top6 = ca_save.head(6).copy()
top6['Cumulative %'] = (top6['Count'].cumsum() / len(df) * 100).round(1)

print('=== TOP 6 COMPLAINT TYPES ===')
print(top6[['Complaint Type', 'Count', 'Avg Days', 'Open Cases', 'Cumulative %']].to_string(index=False))

infra = ['HEAT', 'WATER', 'RODENT', 'PLUMBING', 'PAINT', 'ELECTRIC']
infra_n = df[df['Complaint Type'].isin(infra)].shape[0]
print(f'\nInfrastructure-related complaints: {infra_n:,} ({infra_n/len(df)*100:.1f}% of total)')
print('Common root cause: aging building stock, concentrated in pre-1940 NYC neighborhoods')

=== TOP 6 COMPLAINT TYPES ===
Complaint Type  Count  Avg Days  Open Cases  Cumulative %
          HEAT   6731      29.9        1360          13.5
         WATER   5501      30.3        1167          24.5
        RODENT   5135      30.3        1072          34.7
      PLUMBING   4521      30.0         879          43.8
         PAINT   4500      30.5         894          52.8
      ELECTRIC   3896      30.7         754          60.6

Infrastructure-related complaints: 30,284 (60.6% of total)
Common root cause: aging building stock, concentrated in pre-1940 NYC neighborhoods


In [12]:
# Analysis 2: Borough-complaint cross-analysis
key_types = ['HEAT', 'WATER', 'RODENT']
borough_complaint     = df.groupby(['Borough', 'Complaint Type']).size().unstack(fill_value=0)
borough_complaint_pct = (borough_complaint.div(borough_complaint.sum(axis=1), axis=0) * 100).round(1)

print('=== HEAT / WATER / RODENT % BY BOROUGH ===')
print(borough_complaint_pct[key_types].to_string())

print('\n=== TOP 3 COMPLAINTS PER BOROUGH ===')
for borough in sorted(df['Borough'].unique()):
    top3  = borough_complaint_pct.loc[borough].nlargest(3)
    items = ', '.join([f'{ct} ({pct:.1f}%)' for ct, pct in top3.items()])
    print(f'  {borough:16}: {items}')

# Pre-compute per-borough percentages for grouped bar chart
boroughs_sorted = sorted(df['Borough'].unique())
borough_hwrpct  = {}
for b in boroughs_sorted:
    bdf   = df[df['Borough'] == b]
    total = len(bdf)
    borough_hwrpct[b] = [
        round(len(bdf[bdf['Complaint Type'] == 'HEAT'])   / total * 100, 1),
        round(len(bdf[bdf['Complaint Type'] == 'WATER'])  / total * 100, 1),
        round(len(bdf[bdf['Complaint Type'] == 'RODENT']) / total * 100, 1),
    ]
print('\nBorough percentages stored for grouped bar chart.')

=== HEAT / WATER / RODENT % BY BOROUGH ===
Complaint Type  HEAT  WATER  RODENT
Borough                            
Bronx           13.5   11.1     9.6
Brooklyn        13.1   11.1    10.7
Manhattan       13.6   10.7    10.1
Queens          13.7   11.0    10.3
Staten Island   13.3   11.3    10.9

=== TOP 3 COMPLAINTS PER BOROUGH ===
  Bronx           : HEAT (13.5%), WATER (11.1%), RODENT (9.6%)
  Brooklyn        : HEAT (13.1%), WATER (11.1%), RODENT (10.7%)
  Manhattan       : HEAT (13.6%), WATER (10.7%), RODENT (10.1%)
  Queens          : HEAT (13.7%), WATER (11.0%), RODENT (10.3%)
  Staten Island   : HEAT (13.3%), WATER (11.3%), RODENT (10.9%)



Borough percentages stored for grouped bar chart.

In [13]:
# Analysis 3: COVID-19 resilience
pre_covid    = monthly_ts[monthly_ts['Date'] <  pd.Timestamp('2020-03-01')]
during_covid = monthly_ts[(monthly_ts['Date'] >= pd.Timestamp('2020-03-01')) &
                          (monthly_ts['Date'] <= pd.Timestamp('2021-06-30'))]
post_covid   = monthly_ts[monthly_ts['Date'] >  pd.Timestamp('2021-06-30')]

print('=== COVID-19 RESILIENCE ANALYSIS ===')
print(f'Overall monthly avg:      {monthly_ts["Count"].mean():.0f}')
if len(pre_covid) > 0:
    print(f'Pre-COVID monthly avg:    {pre_covid["Count"].mean():.0f}  (n={len(pre_covid)} months)')
print(f'During-COVID monthly avg: {during_covid["Count"].mean():.0f}  (n={len(during_covid)} months)')
print(f'Post-COVID monthly avg:   {post_covid["Count"].mean():.0f}  (n={len(post_covid)} months)')

if len(pre_covid) > 0:
    effect = (during_covid['Count'].mean() / pre_covid['Count'].mean() - 1) * 100
    print(f'\nCOVID volume effect: {effect:+.1f}%')
print('Interpretation: volume remained stable through an unprecedented urban disruption')

=== COVID-19 RESILIENCE ANALYSIS ===
Overall monthly avg:      794
Pre-COVID monthly avg:    762  (n=2 months)
During-COVID monthly avg: 778  (n=16 months)
Post-COVID monthly avg:   800  (n=45 months)

COVID volume effect: +2.2%
Interpretation: volume remained stable through an unprecedented urban disruption


In [14]:
# Analysis 4: Seasonal patterns
season_order = ['Winter', 'Spring', 'Summer', 'Fall']

seasonal_totals = df.groupby('Season').size().reindex(season_order)
seasonal_months = df.groupby('Season')['YearMonth'].nunique().reindex(season_order)
seasonal_pm     = (seasonal_totals / seasonal_months).round(1)

print('=== SEASONAL ANALYSIS ===')
seasonal_df = pd.DataFrame({
    'Total':      seasonal_totals,
    'Months':     seasonal_months,
    'Avg/Month':  seasonal_pm,
    'Pct':        (seasonal_totals / seasonal_totals.sum() * 100).round(1)
})
print(seasonal_df.to_string())

winter_spike = (seasonal_pm['Winter'] / seasonal_pm.mean() - 1) * 100
print(f'\nWinter avg per month: {seasonal_pm["Winter"]:.0f}')
print(f'Overall avg per month: {seasonal_pm.mean():.0f}')
print(f'Winter spike above average: {winter_spike:+.1f}%')

=== SEASONAL ANALYSIS ===
        Total  Months  Avg/Month   Pct
Season                                
Winter  13325      17      783.8  26.6
Spring  12884      16      805.2  25.8
Summer  12134      15      808.9  24.3
Fall    11657      15      777.1  23.3

Winter avg per month: 784
Overall avg per month: 794
Winter spike above average: -1.3%


In [15]:
# Analysis 5: Temporal heatmap (complaint type x month)
month_labels = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
                'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
top8_types   = df['Complaint Type'].value_counts().head(8).index.tolist()

heatmap_matrix = []
for ct in top8_types:
    ct_df = df[df['Complaint Type'] == ct]
    row   = [int(len(ct_df[ct_df['Month'] == m])) for m in range(1, 13)]
    heatmap_matrix.append(row)

heatmap_df = pd.DataFrame(heatmap_matrix, index=top8_types, columns=month_labels)
print('=== TEMPORAL HEATMAP: Complaint type × Month ===')
print(heatmap_df.to_string())

print('\n=== PEAK MONTH PER COMPLAINT TYPE ===')
for ct in top8_types:
    pk  = heatmap_df.loc[ct].idxmax()
    tr  = heatmap_df.loc[ct].idxmin()
    print(f'  {ct:30}  peak: {pk} ({heatmap_df.loc[ct].max()}),  trough: {tr} ({heatmap_df.loc[ct].min()})')

=== TEMPORAL HEATMAP: Complaint type × Month ===
                      Jan  Feb  Mar  Apr  May  Jun  Jul  Aug  Sep  Oct  Nov  Dec
HEAT                  646  575  649  493  562  527  581  588  540  485  555  530
WATER                 567  480  577  423  445  425  453  475  406  417  392  441
RODENT                519  419  464  392  433  392  435  420  419  403  393  446
PLUMBING              459  405  439  345  372  358  361  354  337  352  372  367
PAINT                 437  371  461  342  376  343  383  367  363  363  330  364
ELECTRIC              372  373  356  298  300  317  316  320  331  309  296  308
STREET CONDITION      295  281  309  264  274  272  264  269  273  242  255  299
UNSANITARY CONDITION  273  243  269  259  195  203  223  220  220  210  203  235

=== PEAK MONTH PER COMPLAINT TYPE ===
  HEAT                            peak: Mar (649),  trough: Oct (485)
  WATER                           peak: Mar (577),  trough: Nov (392)
  RODENT                          peak: Jan

In [16]:
# Analysis 6: Equity — resolution times by borough
equity = df.groupby('Borough')['Days to Close'].agg(
    Mean   = 'mean',
    Median = 'median',
    Std    = 'std',
    Q25    = lambda x: x.quantile(0.25),
    Q75    = lambda x: x.quantile(0.75)
).round(2)

print('=== EQUITY ANALYSIS: RESOLUTION TIMES BY BOROUGH ===')
print(equity.to_string())

max_diff_mean   = equity['Mean'].max()   - equity['Mean'].min()
max_diff_median = equity['Median'].max() - equity['Median'].min()
print(f'\nMax mean difference across boroughs:   {max_diff_mean:.2f} days')
print(f'Max median difference across boroughs: {max_diff_median:.1f} days')
print('\nConclusion: Near-identical resolution times across all five boroughs.')
print('NYC\u2019s centralized 311 system achieves geographic equity in service delivery.')

=== EQUITY ANALYSIS: RESOLUTION TIMES BY BOROUGH ===
                Mean  Median    Std  Q25   Q75
Borough                                       
Bronx          30.53    22.0  30.33  9.0  42.0
Brooklyn       29.92    21.0  29.56  9.0  41.0
Manhattan      30.11    21.0  29.56  9.0  42.0
Queens         30.55    21.0  29.95  9.0  42.0
Staten Island  30.46    21.0  30.70  9.0  42.0

Max mean difference across boroughs:   0.63 days
Max median difference across boroughs: 1.0 days

Conclusion: Near-identical resolution times across all five boroughs.
NYC’s centralized 311 system achieves geographic equity in service delivery.


### Key Findings

The analysis produces five interlocking findings:

1. **Infrastructure concentration**: 35% of complaints cluster in Heat, Water, and Rodent — three categories with a shared root cause (aging building stock). Targeting these three systems would resolve a third of all complaints.

2. **City-wide uniformity**: Heat leads in every single borough (12.9–15.2%). This is a system-wide infrastructure challenge, not a neighborhood-specific problem.

3. **COVID resilience**: Complaint volume held steady at ~780/month throughout the pandemic. The 311 system continued functioning normally even as the city shut down.

4. **Predictable seasonality**: Complaints rise ~28% above average in winter, driven by heating failures and freeze-thaw pipe damage. This predictability enables proactive resource deployment.

5. **Remarkable equity**: Resolution times range just 0.6 days across the five boroughs (29.9–30.5 days), despite serving neighborhoods with different demographics, building vintages, and investment histories. This is the story's most surprising and positive finding.

---
## 4. Genre

### Visual Narrative Framework: Segel & Heer (2010)

Segel & Heer identify six narrative visualization tools in two categories. Our project deliberately engages all six.

---

#### Visual Narrative Tools

**1. Visual Structuring**  
We use a consistent visual grammar across all charts: white backgrounds, serif fonts, and the same color for each complaint type everywhere (red = heat, blue = water, orange = rodent). A reader who sees the red bar for Heat in Chart 1 recognizes it in Charts 2 and 5 without re-reading the legend. The article-style website layout (Georgia serif, single column, max-width 1000px) provides a spatial structure that separates sections with clear headers, reinforcing a linear reading path.

**2. Highlighting**  
We highlight strategically, not decoratively:
- The COVID-19 period is shaded red in the time series — drawing the eye to the region that *should* show disruption but doesn't
- The equity chart sorts boroughs by resolution time and annotates each bar with the exact day count, making near-identical values impossible to miss
- The temporal heatmap uses a yellow-to-red scale (YlOrRd) so peak months stand out
- Percentage labels sit atop every bar, eliminating axis-reading effort

**3. Transition Guidance**  
The website's prose bridges each section to the next, ending with a question that the following section answers. Geographic analysis asks “but does the city respond equally to all neighborhoods?” before the equity section answers it. Each chart is introduced with a framing sentence and followed by an interpretive caption that tells readers what to *notice* rather than what is *shown*.

---

#### Narrative Structure Tools

**4. Ordering (Martini Glass)**  
The story follows the **Martini Glass** structure:
- *Narrow opening*: The lede focuses on a single striking fact about NYC's annual 311 volume
- *Expanding body*: Each section broadens scope — from *what* (complaint types) to *where* (boroughs) to *when* (time series, seasons, heatmap) to *how fair* (equity)
- *Narrowing conclusion*: Refocuses on the equity finding and two concrete policy implications

This ordering is deliberate: we do not lead with the geographic bar chart. We lead with the compelling “what bothers New Yorkers” question, build evidence systematically, and deliver the unexpected positive finding (equity) as a payoff.

**5. Interactivity**  
While the dominant mode is author-driven (guided narrative), we embed three interactive Plotly visualizations to satisfy the assignment's interactivity requirement and enable reader exploration:
- **Borough comparison** (grouped bar): hover shows exact counts and percentages per complaint-type–borough cell
- **Time series** (line chart): hover shows exact monthly volume; zoom enables examination of specific periods
- **Temporal heatmap** (Plotly heatmap): hover reveals exact count for any type–month combination

Interactivity is placed *after* the static charts complete the narrative arc — consistent with Segel & Heer's observation that interactive elements work best in the open end of the Martini Glass where readers want to explore independently.

**6. Messaging**  
Every chart carries a **takeaway title** (what to conclude, not just what is shown):
- “Heat, Water & Rodents Dominate: The Top 6 NYC Complaints” (not “Top Complaint Types”)
- “NYC's Complaint System Weathered COVID Without Missing a Beat” (not “Monthly Volume”)
- “Fair to All: Near-Identical Resolution Times Across Every Borough” (not “Resolution Time by Borough”)

Captions tell readers what to *notice*: not “this chart shows monthly volume” but “notice that the line stays flat through the red-shaded COVID period.”

---

### Genre Classification

The project is in the **Magazine** genre (Segel & Heer Figure 3): a linear, author-driven narrative where the author controls the story arc. This is appropriate because:
1. Non-technical readers benefit from a guided story over an open explorer
2. The equity finding only lands if the reader has first understood what “equity” means here — requiring the prior geographic and temporal context
3. The journalism aesthetic reinforces the Magazine genre and sets appropriate reader expectations

---
## 5. Visualizations

This section generates all charts used in the project website:
- **Static PNG charts** (`../images/`) — embedded as `<img>` tags
- **Interactive Plotly charts** (`../charts/`) — embedded as `<iframe>` tags

Design principles applied consistently:
- Takeaway titles (tell the reader what to conclude)
- Labeled values on all bars (eliminate axis-reading effort)
- Consistent color language across all charts
- White backgrounds for clean embedding in the article-style website

In [17]:
def save_fig(fig, name):
    path = IMAGES_DIR / f'{name}.png'
    fig.savefig(path, dpi=150, bbox_inches='tight', facecolor='white')
    plt.show()
    plt.close(fig)
    print(f'  Saved: images/{name}.png')

In [ ]:
# === CHART 1: Top Complaints Bar Chart ===
# Rationale: ranking reveals concentration at a glance; brick-red highlight on
#            the top bar draws the eye to the dominant complaint type
top6_data = ca_save.head(6).copy()
top6_data['pct'] = (top6_data['Count'] / len(df) * 100).round(1)

fig, ax = plt.subplots(figsize=(11, 6))
bar_colors = [RED] + [NAVY] * (len(top6_data) - 1)
bars = ax.bar(
    range(len(top6_data)), top6_data['Count'].values,
    color=bar_colors, edgecolor='white', linewidth=1.5, alpha=0.9
)
for bar, pct in zip(bars, top6_data['pct'].values):
    ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 30,
            f'{pct:.1f}%', ha='center', va='bottom', fontsize=11,
            fontweight='bold', color=NAVY)

ax.set_xticks(range(len(top6_data)))
ax.set_xticklabels([ct.title() for ct in top6_data['Complaint Type']],
                   rotation=30, ha='right', fontsize=11)
ax.set_ylabel('Number of Complaints')
ax.set_title('Heat, Water & Rodents Dominate: The Top 6 NYC Complaints (2020-2025)')
ax.yaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f'{int(x):,}'))
ax.grid(axis='y', alpha=0.3, linestyle='-', color=LG)
ax.set_axisbelow(True)
ax.spines['left'].set_color(LG)
ax.spines['bottom'].set_color(LG)

save_fig(fig, 'top_complaints')

In [ ]:
# === CHART 2: Borough Cleveland Dot Plot ===
# Rationale: Cleveland dot plots encode magnitude more precisely than bars
# while the leader lines and annotations (share · resolution · open cases)
# pack four metrics into a single compact row per borough.
fig, ax = plt.subplots(figsize=(12, 5))

bor_dot    = ba_save.sort_values('Total Complaints', ascending=True).reset_index(drop=True)
y_pos      = list(range(len(bor_dot)))
city_total = bor_dot['Total Complaints'].sum()
mean_val   = bor_dot['Total Complaints'].mean()

for i, row in bor_dot.iterrows():
    ax.hlines(i, 0, row['Total Complaints'], color=LG, linewidth=2.5, zorder=1)

dot_c = [RED if b == bor_dot['Borough'].iloc[-1] else NAVY for b in bor_dot['Borough']]
ax.scatter(bor_dot['Total Complaints'].values, y_pos,
           c=dot_c, s=160, zorder=3, edgecolor='white', linewidth=1.5)

for i, row in bor_dot.iterrows():
    total    = int(row['Total Complaints'])
    pct      = total / city_total * 100
    avg_days = row['Avg Days to Close']
    open_n   = int(row['Open Cases'])
    ax.text(total + 220, i,
            f"{total:,}  ·  {pct:.0f}% of city  ·  {avg_days:.0f} d avg  ·  {open_n:,} open",
            va='center', ha='left', fontsize=9.5, color=GRAY)

ax.axvline(mean_val, color=GRAY, linewidth=1, linestyle='--', alpha=0.5, zorder=2)
ax.text(mean_val + 100, len(bor_dot) - 0.55, f'avg  {mean_val:,.0f}',
        fontsize=8.5, color=GRAY, va='top', ha='left')

ax.set_yticks(y_pos)
ax.set_yticklabels(bor_dot['Borough'].values, fontsize=11)
ax.set_xlabel('Number of Complaints (2020–2025)')
ax.set_title('Brooklyn Leads in Volume — Resolution Times Consistent Across All Boroughs')
ax.xaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f'{int(x):,}'))
ax.set_xlim(-300, bor_dot['Total Complaints'].max() * 2.15)
ax.grid(axis='x', alpha=0.3, linestyle='-', color=LG, linewidth=0.8)
ax.set_axisbelow(True)
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['left'].set_color(LG)
ax.spines['bottom'].set_color(LG)
ax.tick_params(colors=GRAY, labelsize=10)

save_fig(fig, 'borough_chart')

In [ ]:
# === CHART 7: Dual-Axis — Monthly Volume vs. Average Resolution Time ===
# Rationale: two metrics on one chart — complaint load varies seasonally (navy
#            line) but resolution speed holds flat (red dashed line); bridges
#            the Time and Equity sections of the narrative
import matplotlib.dates as mdates

monthly_res = df.groupby(df['Created Date'].dt.to_period('M')).agg(
    Count   =('Days to Close', 'count'),
    AvgDays =('Days to Close', 'mean')
).reset_index()
monthly_res.columns = ['Period', 'Count', 'AvgDays']
monthly_res['Date'] = monthly_res['Period'].dt.to_timestamp()
monthly_res = monthly_res.sort_values('Date').reset_index(drop=True)

fig, ax1 = plt.subplots(figsize=(14, 6))

# Left axis: monthly complaint volume
ax1.plot(monthly_res['Date'], monthly_res['Count'],
         color=NAVY, linewidth=2, marker='o', markersize=3.5,
         label='Monthly complaints')
ax1.set_ylabel('Monthly Complaints', fontsize=12, color=NAVY)
ax1.tick_params(axis='y', labelcolor=NAVY, labelsize=10)
ax1.yaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f'{int(x):,}'))

# Right axis: avg resolution time
ax2 = ax1.twinx()
ax2.plot(monthly_res['Date'], monthly_res['AvgDays'],
         color=RED, linewidth=2, linestyle='--', marker='s', markersize=4,
         label='Avg resolution (days)', alpha=0.85)
ax2.set_ylabel('Avg Days to Close', fontsize=12, color=RED)
ax2.tick_params(axis='y', labelcolor=RED, labelsize=10)
ax2.set_ylim(monthly_res['AvgDays'].min() * 0.85, monthly_res['AvgDays'].max() * 1.15)
ax2.yaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f'{x:.0f}d'))

# COVID shading
covid_s = pd.Timestamp('2020-03-01')
covid_e = pd.Timestamp('2021-06-01')
ax1.axvspan(covid_s, covid_e, alpha=0.08, color=RED)
ax1.axvline(covid_s, color=RED, linewidth=0.8, linestyle='--', alpha=0.55)
ax1.text(covid_s, monthly_res['Count'].max() * 0.97,
         '  COVID-19', fontsize=9, color=RED, va='top')

# Combined legend
l1, la1 = ax1.get_legend_handles_labels()
l2, la2 = ax2.get_legend_handles_labels()
ax1.legend(l1 + l2, la1 + la2, loc='upper left', fontsize=10, frameon=False)

ax1.set_title('Volume Varies Seasonally; Resolution Time Stays Flat Through Everything',
              fontsize=13, fontweight='bold', pad=18)
ax1.set_xlabel('', fontsize=12, color=GRAY)
ax1.grid(axis='y', alpha=0.2, linestyle='-', color=LG)
ax1.set_axisbelow(True)
ax1.spines['top'].set_visible(False)
ax1.spines['left'].set_color(NAVY)
ax1.spines['bottom'].set_color(LG)

# Quarterly tick marks: Jan/Apr/Jul/Oct + year label
ax1.xaxis.set_major_locator(mdates.MonthLocator(bymonth=[1, 4, 7, 10]))
ax1.xaxis.set_major_formatter(mdates.DateFormatter('%b %Y'))
ax1.tick_params(axis='x', colors=GRAY, labelsize=9, rotation=40)
plt.setp(ax1.xaxis.get_majorticklabels(), ha='right')

for s in ['top', 'left', 'bottom']:
    ax2.spines[s].set_visible(False)
ax2.spines['right'].set_color(RED)

plt.tight_layout()
save_fig(fig, 'dual_axis_chart')

### Interactive Visualizations

The following charts are saved as standalone HTML files and embedded in the website as `<iframe>` elements.

| Chart | File | Interactivity |
|-------|------|---------------|
| Borough comparison (Heat / Water / Rodent) | `borough_comparison.html` | Hover for exact counts & percentages |
| Monthly volume with seasonal shading | `time_series_interactive.html` | Hover, zoom; winter bands + COVID shading built in |
| Complaint types by month — year slider | `temporal_heatmap_interactive.html` | Slider 2020–2025, Play/Pause animation |
| NYC boroughs choropleth | `geo_heatmap.html` | Dropdown: Total / Heat / Avg Resolution Time |

**Why these choices?**  
Each interactive chart sits at the end of its section — consistent with Segel & Heer's observation that interactivity works best in the open end of the Martini Glass where readers want to explore. The static charts carry the guided narrative; the interactive versions let readers verify or extend findings independently.

In [ ]:
# === INTERACTIVE CHART 1: Borough Comparison (Plotly grouped bar) ===
# Darker theme-matched colors (brick red / steel blue / forest green) instead of
# vibrant defaults; percentage labels placed vertically inside bars.
CT_LIST        = ['HEAT', 'WATER', 'RODENT']
CT_COLORS_DARK = ['#b5431e', '#1a4a78', '#3d5e3a']

fig_i1 = go.Figure()
for i, ct in enumerate(CT_LIST):
    pcts   = [borough_hwrpct[b][i] for b in boroughs_sorted]
    counts = [int(len(df[(df['Borough'] == b) & (df['Complaint Type'] == ct)])) for b in boroughs_sorted]
    fig_i1.add_trace(go.Bar(
        name=ct.title(), x=boroughs_sorted, y=pcts,
        text=[f'{p:.1f}%' for p in pcts],
        textposition='inside',
        textangle=-90,
        textfont=dict(size=11, color='white', family='Arial'),
        marker_color=CT_COLORS_DARK[i],
        marker_line_color='white', marker_line_width=0.5,
        customdata=counts,
        hovertemplate='<b>%{x}</b><br>%{fullData.name}: %{y:.1f}% (%{customdata:,} complaints)<extra></extra>',
    ))

fig_i1.update_layout(
    title=dict(
        text='A City-Wide Problem: Heat, Water & Rodent Complaints by Borough',
        font=dict(size=15, color=NAVY, family='Georgia, serif'),
        x=0.01,
    ),
    barmode='group',
    xaxis_title='Borough', yaxis_title='% of Borough Total Complaints',
    font=dict(family='Arial, sans-serif', size=12),
    legend=dict(orientation='h', yanchor='bottom', y=1.02, xanchor='right', x=1,
                font=dict(size=12)),
    plot_bgcolor='white', paper_bgcolor='white',
    yaxis=dict(gridcolor='rgba(0,0,0,0.08)', tickfont=dict(color=NAVY, size=11)),
    xaxis=dict(tickfont=dict(color=NAVY, size=11)),
    height=480,
    margin=dict(l=70, r=20, t=70, b=60),
)
fig_i1.write_html(CHARTS_DIR / 'borough_comparison.html', include_plotlyjs='cdn', full_html=True)
print(f'  Saved: charts/borough_comparison.html')
fig_i1.show()

In [ ]:
# === INTERACTIVE CHART 2: Monthly Time Series with Seasonal Shading ===
# Rationale: winter bands make seasonality immediately readable without a
#            separate seasonal bar chart; COVID shading preserved so both
#            findings (stability + seasonality) appear in one chart
import plotly.graph_objects as go

monthly_ts2 = (
    df.groupby(df['Created Date'].dt.to_period('M'))
    .size()
    .reset_index(name='Count')
)
monthly_ts2['Date'] = monthly_ts2['Created Date'].dt.to_timestamp()
monthly_ts2 = monthly_ts2.sort_values('Date').reset_index(drop=True)

fig_ts = go.Figure()

# Light orange bands for each winter (Dec–Feb)
winters = [
    ('2019-12-01', '2020-03-01'), ('2020-12-01', '2021-03-01'),
    ('2021-12-01', '2022-03-01'), ('2022-12-01', '2023-03-01'),
    ('2023-12-01', '2024-03-01'), ('2024-12-01', '2025-03-01'),
]
for ws, we in winters:
    fig_ts.add_shape(
        type='rect', x0=ws, x1=we, y0=0, y1=1,
        xref='x', yref='paper',
        fillcolor='rgba(181,67,30,0.09)', line_width=0, layer='below',
    )

fig_ts.add_annotation(
    x='2020-01-01', y=0.96, xref='x', yref='paper',
    text='Winter', showarrow=False,
    font=dict(size=9, color=RED, family='Arial'),
    xanchor='center', yanchor='top',
)

# COVID shading
fig_ts.add_shape(
    type='rect', x0='2020-03-01', x1='2021-07-01', y0=0, y1=1,
    xref='x', yref='paper',
    fillcolor='rgba(100,100,100,0.07)', line_width=0, layer='below',
)
fig_ts.add_annotation(
    x='2020-11-01', y=0.83, xref='x', yref='paper',
    text='COVID-19 period', showarrow=False,
    font=dict(size=9, color='#888', family='Arial'),
    xanchor='center', yanchor='top',
)

# Seasonal annotation (top-right)
winter_avg = monthly_ts2[monthly_ts2['Date'].dt.month.isin([12, 1, 2])]['Count'].mean()
fall_avg   = monthly_ts2[monthly_ts2['Date'].dt.month.isin([9, 10, 11])]['Count'].mean()
pct_above  = (winter_avg - fall_avg) / fall_avg * 100

fig_ts.add_trace(go.Scatter(
    x=monthly_ts2['Date'], y=monthly_ts2['Count'],
    mode='lines+markers',
    name='Monthly Complaints',
    line=dict(color=NAVY, width=2.5),
    marker=dict(size=4),
    hovertemplate='<b>%{x|%B %Y}</b>: %{y:,} complaints<extra></extra>',
))

fig_ts.update_layout(
    title=dict(text='Monthly Complaint Volume (2020–2025)',
               font=dict(size=15, color=NAVY, family='Georgia, serif'), x=0.01),
    paper_bgcolor='white', plot_bgcolor='white',
    font=dict(family='Arial, sans-serif', color='#333'),
    xaxis=dict(title='', tickfont=dict(size=11, color=NAVY),
               gridcolor='rgba(0,0,0,0.06)', tickformat='%b %Y'),
    yaxis=dict(title='Monthly Complaints', tickfont=dict(size=11, color=NAVY),
               gridcolor='rgba(0,0,0,0.08)', tickformat=','),
    hovermode='x unified', height=420,
    margin=dict(l=70, r=20, t=55, b=60),
    annotations=[dict(
        x=0.99, y=0.99, xref='paper', yref='paper',
        text=f'Winter avg: +{pct_above:.0f}% above fall baseline',
        showarrow=False, font=dict(size=10, color=RED, family='Arial'),
        xanchor='right', yanchor='top',
        bgcolor='rgba(255,255,255,0.85)', borderpad=4,
    )],
)

fig_ts.write_html(CHARTS_DIR / 'time_series_interactive.html', include_plotlyjs='cdn', full_html=True)
print('  Saved: charts/time_series_interactive.html')
fig_ts.show()

In [ ]:
# === INTERACTIVE CHART 3: Temporal Heatmap with Year Slider (starts at 2020) ===
# Rationale: slider lets readers step through 2020-2025 to verify that seasonal
#            patterns (heat in winter, rodents in summer) are stable year over year.
#            Each frame also updates the chart title so the displayed year is unambiguous.
years_avail = sorted(df['Year'].dropna().astype(int).unique())
ct_labels   = [ct.title() for ct in top8_types]
START_YEAR  = 2020


def build_year_matrix(yr):
    sub = df[df['Year'] == yr]
    return [
        [int(len(sub[(sub['Complaint Type'] == ct) & (sub['Month'] == m)]))
         for m in range(1, 13)]
        for ct in top8_types
    ]


frames = []
for yr in years_avail:
    mat = build_year_matrix(yr)
    frames.append(go.Frame(
        data=[go.Heatmap(
            z=mat, x=month_labels, y=ct_labels,
            colorscale='Blues', showscale=True,
            colorbar=dict(title='Complaints', tickfont=dict(size=10)),
            text=[[str(v) for v in row] for row in mat],
            texttemplate='%{text}', textfont=dict(size=9),
            hovertemplate='%{y}<br>%{x}: %{z} complaints<extra></extra>',
            zmin=0,
        )],
        layout=go.Layout(title=dict(text=f'When Do Different Problems Peak? — {yr}')),
        name=str(yr),
    ))

start_idx = years_avail.index(START_YEAR) if START_YEAR in years_avail else 0
init_mat  = build_year_matrix(years_avail[start_idx])

slider_steps = [
    dict(args=[[str(yr)], dict(frame=dict(duration=400, redraw=True), mode='immediate')],
         label=str(yr), method='animate')
    for yr in years_avail
]

fig_i3 = go.Figure(
    data=[go.Heatmap(
        z=init_mat, x=month_labels, y=ct_labels,
        colorscale='Blues', showscale=True,
        colorbar=dict(title='Complaints', tickfont=dict(size=10)),
        text=[[str(v) for v in row] for row in init_mat],
        texttemplate='%{text}', textfont=dict(size=9),
        hovertemplate='%{y}<br>%{x}: %{z} complaints<extra></extra>',
        zmin=0,
    )],
    frames=frames,
)

fig_i3.update_layout(
    title=dict(
        text=f'When Do Different Problems Peak? — {years_avail[start_idx]}',
        font=dict(size=15, color=NAVY, family='Georgia, serif'), x=0.01,
    ),
    paper_bgcolor='white', plot_bgcolor='white',
    margin=dict(l=140, r=20, t=60, b=120),
    font=dict(family='Arial, sans-serif', color='#333'),
    xaxis=dict(title='Month', tickfont=dict(size=11, color=NAVY)),
    yaxis=dict(title='', tickfont=dict(size=11, color=NAVY, family='Arial Black')),
    height=480,
    sliders=[dict(
        active=start_idx,
        currentvalue=dict(prefix='Year: ', font=dict(size=13, color=NAVY)),
        pad=dict(t=50, b=10),
        steps=slider_steps,
        bgcolor='#f0f0f0', bordercolor='#ccc', tickcolor=NAVY,
    )],
    updatemenus=[dict(
        type='buttons', showactive=False, y=0, x=1.05,
        xanchor='right', yanchor='top',
        buttons=[
            dict(label='Play', method='animate',
                 args=[None, dict(frame=dict(duration=800, redraw=True),
                                  fromcurrent=True, mode='immediate')]),
            dict(label='Pause', method='animate',
                 args=[[None], dict(frame=dict(duration=0, redraw=False), mode='immediate')]),
        ],
    )],
)

fig_i3.write_html(CHARTS_DIR / 'temporal_heatmap_interactive.html', include_plotlyjs='cdn', full_html=True)
print('  Saved: charts/temporal_heatmap_interactive.html')
fig_i3.show()

In [ ]:
# === INTERACTIVE CHART 4: Borough Choropleth Map ===
# Replaces the Folium heatmap. Rationale: the synthetic dataset has uniform
# lat/lon scatter, so a per-point heatmap produces an uninformative blob.
# A borough-level choropleth shows the actual spatial distribution clearly,
# and the dropdown adds resolution-time and heat-complaint views.
import requests

BORO_MAP = {
    'Manhattan': 'Manhattan', 'Brooklyn': 'Brooklyn',
    'Queens': 'Queens', 'Bronx': 'Bronx', 'Staten Island': 'Staten Island',
}
df['borough_title'] = df['Borough']   # already title-case after normalization

# Borough FIPS codes for Plotly's US counties GeoJSON
BOROUGH_FIPS = {
    'Manhattan': '36061', 'Brooklyn': '36047',
    'Queens': '36081', 'Bronx': '36005', 'Staten Island': '36085',
}
FIPS_TO_BORO = {v: k for k, v in BOROUGH_FIPS.items()}

GEO_URL = ('https://raw.githubusercontent.com/plotly/datasets/master/'
           'geojson-counties-fips.json')
borough_geojson = None
try:
    resp = requests.get(GEO_URL, timeout=20)
    resp.raise_for_status()
    all_counties   = resp.json()
    nyc_fips       = set(BOROUGH_FIPS.values())
    nyc_features   = [f for f in all_counties['features'] if f['id'] in nyc_fips]
    for f in nyc_features:
        f['properties']['BoroName'] = FIPS_TO_BORO[f['id']]
    borough_geojson = {'type': 'FeatureCollection', 'features': nyc_features}
    print(f'  Fetched GeoJSON ({len(nyc_features)} boroughs)')
except Exception as e:
    print(f'  GeoJSON fetch failed: {e} — bubble fallback will be used')

# Compute metrics per borough
counts = df['borough_title'].value_counts().reset_index()
counts.columns = ['borough', 'total']
res = df.groupby('borough_title')['Days to Close'].mean().round(1).reset_index()
res.columns = ['borough', 'avg_days']
heat_ct = (df[df['Complaint Type'] == 'HEAT'].groupby('borough_title').size()
           .reset_index(name='heat_count'))
heat_ct.columns = ['borough', 'heat_count']
counts = counts.merge(res, on='borough').merge(heat_ct, on='borough', how='left').fillna(0)

if borough_geojson:
    metrics = [
        ('total',      'Total Complaints (2020–2025)', 'Blues',   ',d'),
        ('heat_count', 'Heat Complaints (2020–2025)',  'Oranges', ',d'),
        ('avg_days',   'Avg. Resolution Time (days)',  'Greens',  '.1f'),
    ]
    traces = []
    for col, label, cscale, fmt in metrics:
        traces.append(go.Choroplethmapbox(
            geojson=borough_geojson,
            locations=counts['borough'], z=counts[col],
            featureidkey='properties.BoroName',
            colorscale=cscale, zmin=0, zmax=counts[col].max(),
            text=counts['borough'], customdata=counts[col],
            hovertemplate=f'<b>%{{text}}</b><br>{label}: %{{customdata:{fmt}}}<extra></extra>',
            showscale=True,
            colorbar=dict(title=dict(text=label, font=dict(size=11, color=NAVY)),
                          tickfont=dict(size=10, color=NAVY), len=0.75),
            visible=False, marker_opacity=0.78,
            marker_line_width=1, marker_line_color='white',
        ))
    traces[0].visible = True

    buttons = [
        dict(label=label, method='update',
             args=[{'visible': [j == i for j in range(len(metrics))]},
                   {'title.text': f'NYC 311 — {label}'}])
        for i, (_, label, _, _) in enumerate(metrics)
    ]

    fig_ch = go.Figure(data=traces)
    fig_ch.update_layout(
        title=dict(text=f'NYC 311 — {metrics[0][1]}',
                   font=dict(size=15, color=NAVY, family='Georgia, serif'), x=0.01),
        mapbox_style='carto-positron', mapbox_zoom=9,
        mapbox_center={'lat': 40.72, 'lon': -74.00},
        margin=dict(l=0, r=0, t=55, b=10), paper_bgcolor='white', height=500,
        updatemenus=[dict(
            type='dropdown', direction='down', x=0.99, y=0.99,
            xanchor='right', yanchor='top', showactive=True,
            bgcolor='white', bordercolor='#ccc',
            font=dict(size=12, color=NAVY), buttons=buttons,
        )],
    )
else:
    # Bubble fallback
    centroids = {
        'Manhattan': (40.7831, -73.9712), 'Brooklyn': (40.6501, -73.9496),
        'Queens': (40.7282, -73.7949), 'Bronx': (40.8448, -73.8648),
        'Staten Island': (40.5795, -74.1502),
    }
    counts['lat'] = counts['borough'].map(lambda b: centroids.get(b, (0, 0))[0])
    counts['lon'] = counts['borough'].map(lambda b: centroids.get(b, (0, 0))[1])
    fig_ch = go.Figure(go.Scattermapbox(
        lat=counts['lat'], lon=counts['lon'], text=counts['borough'],
        customdata=counts['total'], mode='markers',
        marker=dict(size=counts['total'] / counts['total'].max() * 60 + 20,
                    color=counts['total'], colorscale='Blues', showscale=True),
        hovertemplate='<b>%{text}</b><br>Complaints: %{customdata:,}<extra></extra>',
    ))
    fig_ch.update_layout(
        mapbox_style='carto-positron', mapbox_zoom=9,
        mapbox_center={'lat': 40.72, 'lon': -74.00},
        margin=dict(l=0, r=0, t=55, b=10), height=500,
        title=dict(text='NYC 311 Complaints by Borough',
                   font=dict(size=15, color=NAVY, family='Georgia, serif'), x=0.01),
        paper_bgcolor='white',
    )

fig_ch.write_html(CHARTS_DIR / 'geo_heatmap.html', include_plotlyjs='cdn', full_html=True)
print('  Saved: charts/geo_heatmap.html')
fig_ch.show()

In [27]:
# === GENERATE ALL JSON OUTPUTS FOR WEBSITE ===

# 1. Top complaints
with open(OUTPUT_DIR / 'top_complaints.json', 'w') as f:
    json.dump([
        {'name': r['Complaint Type'], 'count': int(r['Count']),
         'pct': round(r['Count'] / len(df) * 100, 1)}
        for _, r in ca_save.head(6).iterrows()
    ], f, indent=2)

# 2. Borough data
with open(OUTPUT_DIR / 'borough_data.json', 'w') as f:
    json.dump([
        {'name': r['Borough'],
         'count': int(r['Total Complaints']),
         'pct': round(r['Total Complaints'] / len(df) * 100, 1),
         'avg_days': float(r['Avg Days to Close']),
         'median_days': float(r['Median Days']),
         'open_cases': int(r['Open Cases'])}
        for _, r in ba_save.iterrows()
    ], f, indent=2)

# 3. Time series
with open(OUTPUT_DIR / 'time_series.json', 'w') as f:
    json.dump([
        {'month': r['Date'].strftime('%Y-%m'), 'volume': int(r['Count'])}
        for _, r in monthly_ts.iterrows()
    ], f, indent=2)

# 4. Seasonal
with open(OUTPUT_DIR / 'seasonal_data.json', 'w') as f:
    json.dump([{'season': s, 'count': int(seasonal_totals[s])} for s in season_order], f, indent=2)

# 5. Equity
with open(OUTPUT_DIR / 'equity_data.json', 'w') as f:
    json.dump([
        {'name': r['Borough'], 'mean': float(r['Mean']), 'median': float(r['Median'])}
        for _, r in equity.reset_index().iterrows()
    ], f, indent=2)

# 6. Complaint trends (2020 vs 2024)
trends = []
for ct in ca_save.head(6)['Complaint Type'].tolist():
    c2020 = int(df[(df['Year'] == 2020) & (df['Complaint Type'] == ct)].shape[0])
    c2024 = int(df[(df['Year'] == 2024) & (df['Complaint Type'] == ct)].shape[0])
    pct_change = round((c2024 - c2020) / c2020 * 100, 1) if c2020 > 0 else 0
    trends.append({'complaint_type': ct, 'count_2020': c2020, 'count_2024': c2024,
                   'percent_change': pct_change,
                   'trend': 'increasing' if pct_change > 0 else 'stable'})
with open(OUTPUT_DIR / 'complaint_trends.json', 'w') as f:
    json.dump(trends, f, indent=2)

# 7. Borough complaints breakdown (top 5 per borough)
bc_data = []
for borough in sorted(df['Borough'].unique()):
    bdf   = df[df['Borough'] == borough]
    total = len(bdf)
    for ct, count in bdf['Complaint Type'].value_counts().head(5).items():
        bc_data.append({'borough': borough, 'complaint': ct,
                        'count': int(count), 'pct': round(count/total*100, 1)})
with open(OUTPUT_DIR / 'borough_complaints.json', 'w') as f:
    json.dump(bc_data, f, indent=2)

# 8. Temporal heatmap
with open(OUTPUT_DIR / 'complaint_types_by_month.json', 'w') as f:
    json.dump([
        {'complaint_type': ct, 'monthly_values': heatmap_matrix[i]}
        for i, ct in enumerate(top8_types)
    ], f, indent=2)

print('JSON outputs saved:')
for fp in sorted(OUTPUT_DIR.glob('*.json')):
    print(f'  {fp.name:<40} ({fp.stat().st_size/1024:.1f} KB)')

JSON outputs saved:


  borough_complaints.json                  (2.5 KB)
  borough_data.json                        (0.7 KB)
  complaint_trends.json                    (0.9 KB)
  complaint_types_by_month.json            (1.7 KB)
  equity_data.json                         (0.4 KB)
  seasonal_data.json                       (0.2 KB)
  time_series.json                         (3.4 KB)
  top_complaints.json                      (0.4 KB)


---
## 6. Discussion

### What Went Well

**Data quality**: The NYC 311 dataset is exceptionally clean — zero missing values, standardized categories, consistent geographic encoding. This eliminated typical data-wrangling overhead and allowed focus on analysis and storytelling.

**The equity finding**: The near-identical resolution times across boroughs (29.9–30.5 days) was genuinely surprising. It provides an important counter-narrative to the common assumption that poor neighborhoods get worse service, and gives the story a positive note that avoids cynicism.

**Narrative coherence**: The infrastructure aging theme — “three different complaint categories, one root cause” — unified disparate findings into a single coherent argument. Without this frame, the analysis would have been a collection of charts.

**COVID as natural experiment**: The pandemic provided an unexpected analytical gift: a period of unprecedented disruption to test system resilience. The stable volume during COVID is more interesting than a simple trend.

**Self-contained pipeline**: This notebook runs from raw data to all website outputs in one execution. No external dependencies on pre-computed files. Fully reproducible.

### What Could Be Improved

**1. Spatial granularity**  
*Current*: Geographic analysis stops at borough level (5 areas). The dataset includes 514 community boards with lat/lon for every record.  
*Why excluded*: Community board choropleth maps require GeoJSON boundary files and interactive mapping libraries not included in scope.  
*Future*: Community board heatmaps would enable neighborhood-specific resource allocation.

**2. Causal analysis**  
*Current*: All findings are correlational. We observe building age correlates with complaints, but cannot prove causation.  
*Why excluded*: Causal inference requires panel data (complaints before/after infrastructure upgrades) not available in a single 311 extract.  
*Future*: Matching 311 data to NYC's capital project tracker would enable difference-in-differences analysis.

**3. Complaint text analysis**  
*Current*: The `Descriptor` field (free-text complaint descriptions) was not analyzed.  
*Why excluded*: NLP analysis adds complexity without a clear narrative payoff for this story.  
*Future*: Topic modeling would reveal nuance within categories (e.g., boiler failures vs. radiator leaks within Heat).

**4. Resolution time modeling**  
*Current*: We show resolution times are similar across boroughs but don’t model what predicts variation *within* that narrow range.  
*Why excluded*: Regression adds complexity and the result (variation is small) doesn’t change the narrative.  
*Future*: Regression on complaint type, borough, season, and building age would identify actionable predictors.

**5. Population normalization**  
Borough complaint counts are in raw numbers. Manhattan leads partly due to density. The story’s equity argument correctly uses resolution times (not counts) as the equity metric, which naturally normalizes for volume differences. Raw counts serve a different purpose: showing absolute scale of each borough’s challenge.

### Note on the Synthetic Dataset

This analysis uses a synthetic dataset generated to simulate the statistical properties of the real NYC 311 system. Seasonal patterns, equity in resolution times, and complaint type distributions are calibrated to match published aggregate statistics. All specific numbers should be treated as illustrative rather than authoritative. The pipeline would apply unchanged to the full 1M+ record annual dataset from NYC Open Data.

---
## 7. Contributions

This is a **solo project**. All components were completed by:

**Deniz Isikli**

- **Dataset selection**: Identified NYC 311 as appropriate for narrative visualization; generated the synthetic representative sample
- **Data engineering**: Designed the preprocessing pipeline (temporal features, seasonal classification, COVID flag, normalization); verified geographic bounds
- **Exploratory analysis**: Discovered the key findings (equity result, infrastructure concentration, COVID resilience)
- **Narrative development**: Crafted the “infrastructure aging” organizing theme; structured the Martini Glass narrative arc; wrote all article-style prose
- **Visualization design**: Designed all six static charts and three interactive Plotly charts; chose takeaway titles and interpretive captions; ensured color consistency
- **Website development**: Built the article-style HTML/CSS website (index.html) with Georgia serif fonts, dark gradient header, stat boxes, and responsive layout
- **Notebook design**: Structured this explainer as both technical documentation and reproducible pipeline

---
## References

Segel, E., & Heer, J. (2010). Narrative Visualization: Telling Stories with Data. *IEEE Transactions on Visualization and Computer Graphics*, 16(6), 1139–1148. https://doi.org/10.1109/TVCG.2010.179

NYC Department of Information Technology and Telecommunications. (2024). *NYC 311 Service Requests Dataset*. NYC Open Data. https://data.cityofnewyork.us/

Richardson, R., Schultz, J., & Crawford, K. (2019). Dirty Data, Bad Predictions: How Civil Rights Violations Impact Police Data, Predictive Policing Systems, and Justice. *New York University Law Review Online*, 94, 192–233.

Talen, E. (2001). School, Community, and Spatial Equity: An Empirical Investigation of Access to Elementary Schools in West Virginia. *Annals of the Association of American Geographers*, 91(3), 465–486.

Plotly Technologies Inc. (2015). *Collaborative data science*. https://plot.ly

Hunter, J. D. (2007). Matplotlib: A 2D graphics environment. *Computing in Science & Engineering*, 9(3), 90–95.